### ISSUE TITLE
Support Builder Mode in Grafana 

### ISSUE THREAD

**Issue by tenmozes** [VictoriaMetrics/victorialogs-datasource#48]
We need to add support for autocompletion for labels in Builder mode in Grafana 

the same feature in Loki 

![builder-ac](https://github.com/user-attachments/assets/8d436704-e3e2-41bd-bbb5-8db810c6a115)

**Comment by Loori-R** [comment]
**Roadmap for Future Improvements:**

- [ ] **Support for the NOT operator**. Example expressions: `_time:5m error NOT buggy_app` and `_time:5m log.level:error !(buggy_app OR foobar)`.
- [ ] **Support for creating nested queries directly from the UI builder**. Currently, it can parse nested queries but does not allow users to create them.
- [ ] **Support for using pipes in queries**. The list of available pipes can be found [here](https://docs.victoriametrics.com/victorialogs/logsql/#filters).
- [ ] **Support for various LogsQL filters**. The complete list of filters is available [here](https://docs.victoriametrics.com/victorialogs/logsql/#filters).
- [ ] **Support for expressions where application identifiers (e.g., `buggy_app` and `foobar`) are stored in a specific field**. For example: `_time:5m log.level:error !(app:buggy_app OR app:foobar)`

**Comment by archef2000** [comment]
Is there a plan for when pipes (more specifically the format pipe) will be implemented? Is a PR for this feature welcome? I come from loki, but like the ease of HA in VM and was already using it for metrics so I also switched to VL.

**Comment by Loori-R** [comment]
Hi, @archef2000! Pipe support is planned, but not in the near future. We’d be happy to review a PR if someone is up for it!

**Comment by archef2000** [comment]
First as Queries are also supported after a pipe I would combine filters and operations and just select a operation like ["label filter", "matcher", etc].

Second I have some problems testing my changes. What is your workflow for testing frontend changes?

**Comment by Loori-R** [comment]
1. I agree, but we need to be careful here — some pipe operations allow parts of the phrase to be omitted. For example, `sort` can be written as `sort by`, `| sort`, or `| order by`, so we need to handle those variations properly.

2. For testing, we have test files like `src/components/QueryEditor/QueryBuilder/utils/parsing.test.ts`. You can add your own tests there if needed. If necessary, feel free to add your own test files as well.

**Comment by archef2000** [comment]
No I mean how are you quickly testing UI changes in Grafana, as It tkes me a lot of time building the ui everytime I change something and restart grafana.

**Comment by archef2000** [comment]
I already tried that, but I just get `{"tinylr":"Welcome","version":"1.1.1"}` how can I then see the UI?

**Comment by Loori-R** [comment]
[Full instructions on GitHub](https://github.com/VictoriaMetrics/victorialogs-datasource#getting-started-development)

1. **Start the frontend dev server:**

   ```bash
   npm run dev
   ```

2. **Update `grafana.ini`:**

   ```ini
   [paths]
   plugins = ../../../WebstormProjects/grafana-logs-datasource/plugins

   [plugins]
   allow_loading_unsigned_plugins = victoriametrics-logs-datasource
   ```

3. **Start Grafana:**

   ```bash
   ./bin/grafana-server web
   ```

4. **Open Grafana in the browser with the developer console enabled.**
   If changes don't appear — do a hard reload with cache clear (`Ctrl+Shift+R` or `Cmd+Shift+R`).

**Comment by archef2000** [comment]
Thanks. Didn't notice that the dev command also updates the plugin

**Comment by archef2000** [comment]
Is there a performance difference between `field1: "test" AND fiel2: "test"` and `field1: "test" | fiel2: "test"`

**Comment by Loori-R** [comment]
> Is there a performance difference between `field1: "test" AND fiel2: "test"` and `field1: "test" | fiel2: "test"`

There doesn't seem to be any performance difference

**Comment by Loori-R** [comment]
Ideally, the query should remain unchanged when converting from Editor to Builder. Minor technical changes are fine if the logic and result stay the same. Just to clarify — formatting or logic changes?

**Comment by archef2000** [comment]
I mean:
* `field1: "test" AND fiel2: "test"` to `field1: "test" | fiel2: "test"`
* `{label!~"v1|...|vN"}` to `{label not_in (v1,...,vN)}`

**Comment by archef2000** [comment]
@Loori-R May you be able to provide some complex test queries with edge cases to test my parser?

**Comment by Loori-R** [comment]
@archef2000, Here are some examples:
```
_time:30m ("cannot open file" OR i("segmentation fault")) -debug | stream_context before 3 after 2

_time:[2025-05-01T00:00:00Z, 2025-05-02T00:00:00Z) "User "admin" logged in"

_time:2h (client_ip:ipv4_range("10.0.0.0/8") OR client_ip:ipv4_range("192.168.1.100"))

_time:4h request.duration:range(4.2, Inf] response_size_bytes:range[1_000_000, Inf)

_time:10m len_range(0, 32) contains_any(error, "warn")

_time:24h user_id:value_type(uint64) | stats count_uniq(user_id) uniq_users

_time:1d {app="frontend"} | stats by (user) count() frontend_hits | join by (user) (_time:1d {app="backend"} | stats by (user) avg(latency_ms) backend_avg_lat) inner

_time:6h | stats by (_time:10m) count() total, count() if (log.level:error) errors | math errors / total as error_rate

_time:5m user_id:in(_time:1d path:admin | fields user_id) | first 50 by (_time desc)

_time:30m error | union (_time:2h panic) | uniq by (_stream)

kubernetes.pod_namespace:in($namespace) kubernetes.pod_name:in($pod) kubernetes.container_name:in($container)
| format if (kubernetes.pod_namespace:"") "unknown" as kubernetes.pod_namespace
| stats by (kubernetes.pod_namespace) count() total
| sort by(total) desc 
| limit 5
```

@hagen1778, @dmitryk-dk — feel free to add more if you have any other ideas or edge cases in mind.

**Comment by archef2000** [comment]
A little update: Most operations are working.

![Image](https://github.com/user-attachments/assets/98c7aeb7-251b-4419-95ff-7bd88663d4fc)

**Comment by Loori-R** [comment]
> Does this exact query work with the quotes?
> `_time:[2025-05-01T00:00:00Z, 2025-05-02T00:00:00Z) "User "admin" logged in"`

My mistake: this part needs to be escaped — `"User "admin" logged in"` → `"User \"admin\" logged in"`

<hr/>

> Is the desc property per field or for all fields?
> `_time:5m user_id:in(_time:1d path:admin | fields user_id) | first 50 by (_time desc)`

The `desc` keyword applies only to the field it directly follows. In your example, only `_time` is sorted in descending order. It does not affect other fields unless explicitly specified.
As stated in the documentation:

> `| first N by (fields)` pipe returns the first N logs from a query after sorting them by the given fields.

See more in the official docs: [VictoriaLogs `first` pipe](https://docs.victoriametrics.com/victorialogs/logsql/#first-pipe)

**Comment by Loori-R** [comment]
In LogSQL, case-insensitive search uses `i("...")` - without the `=`. See [this docs](https://docs.victoriametrics.com/victorialogs/logsql/#case-insensitive-filter).
For example, this is correct:

```
file:i("oauth2-proxy")
```

Using `=i("...")` will cause an error. Could you clarify what exactly you're trying to search for?

**Comment by archef2000** [comment]
Ah ok. I just tested all possible options for all operations and came across the exact + case insensitive combination.

**Comment by archef2000** [comment]
I am currently going through all the query examples to check everything.
Is this a valid query? (specifically the unroll part)
`
_time:5m | unpack_words from _msg as words | unroll words | top 5 (words)
`

**Comment by dmitryk-dk** [comment]
> _time:5m | unpack_words from _msg as words | unroll words | top 5 (words)
Hi @archef2000 , this is valid query

<img width="1799" alt="Image" src="https://github.com/user-attachments/assets/dc1ce33d-b276-4ffb-b0e7-82a5077c1511" />

**Comment by archef2000** [comment]
So the words is for the unroll operation. Are multiple fields allowed seperated by a comma?

**Comment by Loori-R** [comment]
> So the words is for the unroll operation. Are multiple fields allowed seperated by a comma?

Yes, multiple fields separated by commas are supported in the `unroll` operation. This feature was added in [v1.12.0](https://docs.victoriametrics.com/victorialogs/changelog/#v1120).

For example:
`_time:5m | unroll timestamp, value, _msg`

You can also try and validate queries at: http://play-vmlogs.victoriametrics.com/select/vmui/

**Comment by archef2000** [comment]
So that query in the docs is wrong and there are many features like unroll
https://docs.victoriametrics.com/victorialogs/logsql/#comments
`
cannot parse query [* | stats by (_stream) logs]: cannot parse 'stats' pipe: unknown stats func "logs"; context: [* | stats by (_stream) logs]
`

**Comment by Loori-R** [comment]
> So that query in the docs is wrong and there are many features like in the unroll operation that are missing. https://docs.victoriametrics.com/victorialogs/logsql/#comments `cannot parse query [* | stats by (_stream) logs]: cannot parse 'stats' pipe: unknown stats func "logs"; context: [* | stats by (_stream) logs]`

Yes, the query in the docs does seem incorrect. Thanks for spotting it!
This part:

```logsql
| stats by (_stream) logs
```

is missing an aggregate function. It should likely be something like:

```logsql
| stats by (_stream) count() logs
```

Looks like a simple typo. PR with a fix: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/9272.

**Comment by arturminchukov** [comment]
Query builder was added in release [v0.27.0](https://github.com/VictoriaMetrics/victorialogs-datasource/releases/tag/v0.27.0). So I'm closing this issue. If you notice some bugs or would like to suggest some enhancements, please create a new issue. Read more about query builder - https://docs.victoriametrics.com/victorialogs/integrations/grafana/#query-builder
